### DISCUSSION TITLE
Paraxial calculations

### DISCUSSION THREAD

**Discussion by dibyendumajumdar** [mjhoptics/ray-optics#44] [General]
Hi Mike,

I am trying to understand the code in `parax/firstorder.py`. In particular the first section of `compute_first_order`. In the literature I have I don't see this part described; I guess it is to determine the initial set of rays in the paraxial region? Please let me know any references to literature regarding this.

Thanks and Regards
Dibyendu

**Comment by mjhoptics** [comment]
The idea in `compute_first_order` is to use paraxial matrix optics to calculate the entrance pupil size and location and other first order quantities. It turns out that the element of the matrices (sometimes called ABCD matrices) can be obtained directly by tracing 2 linearly independent paraxial rays. The so-called p-ray has unit height and zero slope at the first surface and the second, the q-ray, has zero height and unit slope at the first surface. By tracing these rays through the system, a paraxial transfer matrix can be constructed for transfer from the 1st surface to the kth surface directly in terms of the p and q ray data at the kth surface. 

I use this approach to build transfer matrices from the first to last surfaces and from the first surface to the stop surface, to calcualte the paraxial entrance pupil location. The transfer from 1st to last surfaces is used for focal length, principal planes, magnification, etc.

There is some info on the web about paraxial transfer matrices, ABCD matrices and the like. The details of what I use are in unpublished notes by Doug Sinclair, from when I was an optics student at the Inst of Optics (a long time ago).  I haven't found this material presented elsewhere although it was a common thread of instruction at the Institute for several decades.

I'll add some comments to the function to make it a little more obvious what is being done and why.

**Comment by dibyendumajumdar** [comment]
Okay thank you - your detailed explanations are much appreciated!
I did find some info on the ABCD matrices in Optical System Design by Kingslake - hopefully I can debug the code and that way gain better understanding.
